File: mobile/lib/net/spam.dart
class SpamWatch {
  SpamWatch();

  static const banSec = 30;
  static const burstN = 4;
  static const burstSec = 2.0;
  static const repeatN = 3;
  static const repeatSec = 10.0;
  static const rapidN = 3;
  static const rapidGap = 0.4;

  final _events = <String, List<(double, String)>>{};

  bool note(String who, String text) {
    if (who.isEmpty) {
      return false;
    }
    final line = text.toLowerCase().split(RegExp(r'\s+')).where((p) => p.isNotEmpty).join(' ');
    if (line.isEmpty) {
      return false;
    }
    final now = DateTime.now().millisecondsSinceEpoch / 1000.0;
    final q = _events.putIfAbsent(who, () => <(double, String)>[]);
    q.add((now, line));
    q.removeWhere((e) => now - e.$1 > repeatSec);
    if (q.where((e) => now - e.$1 <= burstSec).length >= burstN) {
      return true;
    }
    if (q.where((e) => e.$2 == line).length >= repeatN) {
      return true;
    }
    if (q.length >= rapidN) {
      final times = q.sublist(q.length - rapidN).map((e) => e.$1).toList();
      var rapid = true;
      for (var i = 1; i < times.length; i++) {
        if (times[i] - times[i - 1] > rapidGap) {
          rapid = false;
          break;
        }
      }
      if (rapid) {
        return true;
      }
    }
    return false;
  }

  void forget(String who) {
    _events.remove(who);
  }
}

File: ephemeral_chat/spam.py
"""Host-side spam detector. RAM only."""

from __future__ import annotations

import time
from collections import defaultdict, deque

BAN_SEC = 30
BURST_N = 4
BURST_SEC = 2.0
REPEAT_N = 3
REPEAT_SEC = 10.0
RAPID_N = 3
RAPID_GAP = 0.4


def _norm(text: str) -> str:
    return " ".join(text.lower().split())


class SpamWatch:
    def __init__(self) -> None:
        self._events: dict[str, deque[tuple[float, str]]] = defaultdict(deque)

    def note(self, who: str, text: str, now: float | None = None) -> bool:
        """Record a chat line. True if this sender is spamming."""
        if not who:
            return False
        now = time.time() if now is None else now
        line = _norm(text)
        if not line:
            return False
        q = self._events[who]
        q.append((now, line))
        while q and now - q[0][0] > REPEAT_SEC:
            q.popleft()
        burst = sum(1 for ts, _ in q if now - ts <= BURST_SEC)
        if burst >= BURST_N:
            return True
        if sum(1 for _, body in q if body == line) >= REPEAT_N:
            return True
        times = [ts for ts, _ in q][-RAPID_N:]
        if len(times) >= RAPID_N and all(times[i] - times[i - 1] <= RAPID_GAP for i in range(1, len(times))):
            return True
        return False

    def forget(self, who: str) -> None:
        self._events.pop(who, None)
